File: server/app/utils/audit_ddl_trail.py
"""Bug-99: DDL 审计链.

设计:
  - 记录 DDL 操作 (CREATE / ALTER / DROP) 到追加日志
  - 每条记录包含前一记录的 SHA-256 哈希 (链式防篡改)
  - 启动时可校验整条链完整性
  - 支持内存索引按表名 / 操作者 / 时间范围查询
  - 持久化到 JSONL, 异常恢复
"""

import hashlib
import json
import logging
import os
import threading
import time
from collections import deque
from dataclasses import asdict, dataclass, field
from enum import StrEnum

logger = logging.getLogger(__name__)

GENESIS_HASH = "0" * 64


class DdlOp(StrEnum):
    CREATE = "CREATE"
    ALTER = "ALTER"
    DROP = "DROP"
    RENAME = "RENAME"
    TRUNCATE = "TRUNCATE"
    INDEX = "INDEX"


@dataclass
class DdlEntry:
    op: str
    obj_type: str  # TABLE / INDEX / VIEW / ...
    obj_name: str
    actor: str
    ts: float
    sql: str
    prev_hash: str = GENESIS_HASH
    hash: str = ""
    extra: dict = field(default_factory=dict)

    def compute_hash(self) -> str:
        body = f"{self.op}|{self.obj_type}|{self.obj_name}|{self.actor}|{self.ts}|{self.sql}|{self.prev_hash}"
        h = hashlib.sha256(body.encode("utf-8")).hexdigest()
        return h

    def to_dict(self) -> dict:
        d = asdict(self)
        return d

    @classmethod
    def from_dict(cls, d: dict) -> "DdlEntry":
        return cls(**d)


class DdlAuditTrail:
    """DDL 审计链: 追加 + 哈希链 + 校验."""

    def __init__(self, log_path: str = "", max_in_mem: int = 5000):
        self._lock = threading.Lock()
        self._log_path = log_path
        self._entries: deque[DdlEntry] = deque(maxlen=max_in_mem)
        self._tail_hash: str = GENESIS_HASH
        if log_path and os.path.exists(log_path):
            self._load_from_file()

    def _load_from_file(self) -> None:
        try:
            with open(self._log_path, encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        d = json.loads(line)
                        e = DdlEntry.from_dict(d)
                        self._entries.append(e)
                        self._tail_hash = e.hash
                    except Exception:
                        continue
        except OSError as e:
            logger.warning("ddl_audit load failed: %s", e)

    def _append_to_file(self, entry: DdlEntry) -> None:
        if not self._log_path:
            return
        try:
            with open(self._log_path, "a", encoding="utf-8") as f:
                f.write(json.dumps(entry.to_dict(), ensure_ascii=False) + "\n")
        except OSError as e:
            logger.warning("ddl_audit append failed: %s", e)

    def record(
        self,
        op: str,
        obj_type: str,
        obj_name: str,
        actor: str,
        sql: str,
        extra: dict | None = None,
    ) -> DdlEntry:
        op = str(op).upper()
        with self._lock:
            entry = DdlEntry(
                op=op,
                obj_type=obj_type,
                obj_name=obj_name,
                actor=actor,
                ts=time.time(),
                sql=sql,
                prev_hash=self._tail_hash,
                extra=extra or {},
            )
            entry.hash = entry.compute_hash()
            self._entries.append(entry)
            self._tail_hash = entry.hash
            self._append_to_file(entry)
            logger.info("ddl_audit: op=%s obj=%s.%s actor=%s", op, obj_type, obj_name, actor)
            return entry

    def verify(self) -> dict[str, object]:
        """校验整条链, 返回 (ok, broken_at, total)."""
        with self._lock:
            prev = GENESIS_HASH
            broken_at: int | None = None
            for idx, e in enumerate(self._entries):
                if e.prev_hash != prev:
                    broken_at = idx
                    break
                recomputed = e.compute_hash()
                if recomputed != e.hash:
                    broken_at = idx
                    break
                prev = e.hash
            total = len(self._entries)
            return {
                "ok": broken_at is None,
                "broken_at": broken_at,
                "total": total,
                "tail_hash": self._tail_hash if self._entries else GENESIS_HASH,
            }

    def list_recent(self, n: int = 50, obj_name: str | None = None) -> list[DdlEntry]:
        with self._lock:
            arr = list(self._entries)
        if obj_name:
            arr = [e for e in arr if e.obj_name == obj_name]
        return arr[-n:]

    def query(
        self,
        obj_name: str | None = None,
        actor: str | None = None,
        op: str | None = None,
        since_ts: float = 0.0,
        limit: int = 200,
    ) -> list[DdlEntry]:
        with self._lock:
            arr = list(self._entries)
        out: list[DdlEntry] = []
        for e in arr:
            if obj_name and e.obj_name != obj_name:
                continue
            if actor and e.actor != actor:
                continue
            if op and e.op != op.upper():
                continue
            if e.ts < since_ts:
                continue
            out.append(e)
            if len(out) >= limit:
                break
        return out

    def clear(self) -> None:
        with self._lock:
            self._entries.clear()
            self._tail_hash = GENESIS_HASH
        if self._log_path and os.path.exists(self._log_path):
            try:
                os.remove(self._log_path)
            except OSError:
                logger.warning("Caught unexpected exception")

    def stats(self) -> dict:
        with self._lock:
            total = len(self._entries)
            by_op: dict[str, int] = {}
            for e in self._entries:
                by_op[e.op] = by_op.get(e.op, 0) + 1
            return {
                "total": total,
                "by_op": by_op,
                "tail_hash": self._tail_hash,
                "log_path": self._log_path,
            }

    def set_log_path(self, p: str) -> None:
        with self._lock:
            self._log_path = p


# 全局单例
ddl_audit = DdlAuditTrail()

File: apps/api/src/utils/audit-ddl-trail.ts
/**
 * DDL 审计追踪链.
 *
 * 设计:
 *   - 记录 DDL 操作 (CREATE / ALTER / DROP / RENAME / TRUNCATE / INDEX) 到追加日志
 *   - 每条记录包含前一记录的 SHA-256 哈希 (链式防篡改)
 *   - 启动时可校验整条链完整性
 *   - 支持内存索引按表名 / 操作者 / 时间范围查询
 *   - 持久化到 JSONL, 异常恢复
 *
 * 参考: git show 3ee96cf0:server/app/utils/audit_ddl_trail.py
 */

import { createHash } from 'node:crypto'
import { appendFileSync, existsSync, readFileSync, rmSync } from 'node:fs'

/** 创世哈希. */
export const GENESIS_HASH = '0'.repeat(64)

/** DDL 操作类型. */
export enum DdlOp {
  CREATE = 'CREATE',
  ALTER = 'ALTER',
  DROP = 'DROP',
  RENAME = 'RENAME',
  TRUNCATE = 'TRUNCATE',
  INDEX = 'INDEX',
}

/** DDL 审计条目. */
export interface DdlEntry {
  op: string
  /** 对象类型: TABLE / INDEX / VIEW ... */
  objType: string
  objName: string
  actor: string
  ts: number
  sql: string
  prevHash: string
  hash: string
  extra: Record<string, unknown>
}

/** 校验结果. */
export interface DdlVerifyResult {
  ok: boolean
  /** 断点位置, null 表示无断点 */
  brokenAt: number | null
  total: number
  tailHash: string
}

/** 计算条目哈希. */
function computeHash(entry: DdlEntry): string {
  const body = `${entry.op}|${entry.objType}|${entry.objName}|${entry.actor}|${entry.ts}|${entry.sql}|${entry.prevHash}`
  return createHash('sha256').update(body, 'utf8').digest('hex')
}

/**
 * DDL 审计链: 追加 + 哈希链 + 校验.
 */
export class DdlAuditTrail {
  private readonly maxInMem: number
  private logPath: string
  private readonly entries: DdlEntry[] = []
  private tailHash = GENESIS_HASH

  constructor(logPath = '', maxInMem = 5000) {
    this.logPath = logPath
    this.maxInMem = maxInMem
    if (logPath && existsSync(logPath)) this.loadFromFile()
  }

  private loadFromFile(): void {
    try {
      const text = readFileSync(this.logPath, 'utf8')
      for (const line of text.split('\n')) {
        const trimmed = line.trim()
        if (!trimmed) continue
        try {
          const d = JSON.parse(trimmed) as DdlEntry
          this.entries.push(d)
          this.tailHash = d.hash
        } catch {
          continue
        }
      }
      // 超出内存上限: 保留最新的
      if (this.entries.length > this.maxInMem) {
        this.entries.splice(0, this.entries.length - this.maxInMem)
      }
    } catch {
      /* 读取失败忽略 */
    }
  }

  private appendToFile(entry: DdlEntry): void {
    if (!this.logPath) return
    try {
      appendFileSync(this.logPath, JSON.stringify(entry) + '\n', 'utf8')
    } catch {
      /* 写入失败忽略 */
    }
  }

  /** 记录一条 DDL. */
  record(
    op: string,
    objType: string,
    objName: string,
    actor: string,
    sql: string,
    extra?: Record<string, unknown>,
  ): DdlEntry {
    const opUpper = op.toUpperCase()
    const entry: DdlEntry = {
      op: opUpper,
      objType,
      objName,
      actor,
      ts: Date.now() / 1000,
      sql,
      prevHash: this.tailHash,
      hash: '',
      extra: extra ?? {},
    }
    entry.hash = computeHash(entry)
    this.entries.push(entry)
    if (this.entries.length > this.maxInMem) this.entries.shift()
    this.tailHash = entry.hash
    this.appendToFile(entry)
    return entry
  }

  /** 校验整条链. */
  verify(): DdlVerifyResult {
    let prev = GENESIS_HASH
    let brokenAt: number | null = null
    for (let idx = 0; idx < this.entries.length; idx++) {
      const e = this.entries[idx]!
      if (e.prevHash !== prev) {
        brokenAt = idx
        break
      }
      const recomputed = computeHash(e)
      if (recomputed !== e.hash) {
        brokenAt = idx
        break
      }
      prev = e.hash
    }
    return {
      ok: brokenAt === null,
      brokenAt,
      total: this.entries.length,
      tailHash: this.entries.length > 0 ? this.tailHash : GENESIS_HASH,
    }
  }

  /** 获取最近 N 条. */
  listRecent(n = 50, objName?: string): DdlEntry[] {
    let arr = this.entries
    if (objName) arr = arr.filter((e) => e.objName === objName)
    return arr.slice(-n)
  }

  /** 按条件查询. */
  query(
    opts: {
      objName?: string
      actor?: string
      op?: string
      sinceTs?: number
      limit?: number
    } = {},
  ): DdlEntry[] {
    const { objName, actor, op, sinceTs = 0, limit = 200 } = opts
    const opUpper = op?.toUpperCase()
    const out: DdlEntry[] = []
    for (const e of this.entries) {
      if (objName && e.objName !== objName) continue
      if (actor && e.actor !== actor) continue
      if (opUpper && e.op !== opUpper) continue
      if (e.ts < sinceTs) continue
      out.push(e)
      if (out.length >= limit) break
    }
    return out
  }

  /** 清空 (同时删除持久化文件). */
  clear(): void {
    this.entries.length = 0
    this.tailHash = GENESIS_HASH
    if (this.logPath && existsSync(this.logPath)) {
      try {
        rmSync(this.logPath)
      } catch {
        /* ignore */
      }
    }
  }

  /** 统计. */
  stats(): { total: number; byOp: Record<string, number>; tailHash: string; logPath: string } {
    const byOp: Record<string, number> = {}
    for (const e of this.entries) byOp[e.op] = (byOp[e.op] ?? 0) + 1
    return {
      total: this.entries.length,
      byOp,
      tailHash: this.tailHash,
      logPath: this.logPath,
    }
  }

  /** 设置日志路径. */
  setLogPath(p: string): void {
    this.logPath = p
  }
}

/** 全局单例. */
export const ddlAudit = new DdlAuditTrail()
